include/zephyr/drivers/stepper.h

```diff
@@ -0,0 +1,241 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_STEPPER_H
+#define ZEPHYR_INCLUDE_STEPPER_H
+
+/** \addtogroup stepper
+ *  @{
+ */
+
+/**
+ * @file
+ * @brief Public API for Stepper Motor Controller
+ *
+ */
+
+#include <zephyr/kernel.h>
+#include <errno.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Stepper Motor micro step resolution options
+ */
+enum micro_step_resolution {
+	STEPPER_FULL_STEP,
+	STEPPER_MICRO_STEP_2,
+	STEPPER_MICRO_STEP_4,
+	STEPPER_MICRO_STEP_8,
+	STEPPER_MICRO_STEP_16,
+	STEPPER_MICRO_STEP_32,
+	STEPPER_MICRO_STEP_64,
+	STEPPER_MICRO_STEP_128,
+	STEPPER_MICRO_STEP_256,
+};
+
+/**
+ * @brief Attibutes to set/get as per the functionality offered by the stepper controller
+ */
+enum stepper_attribute {
+	/** read/write actual a.k.a reference position in the stepper driver */
+	STEPPER_ACTUAL_POSITION,
+	/** read/write target a.k.a absolute position in the stepper driver */
+	STEPPER_TARGET_POSITION,
+	/** setting this attribute should activate constant velocity mode with the given velocity */
+	STEPPER_CONSTANT_VELOCITY,
+	/** start ramp velocity, for controllers which provide ramping functionality */
+	STEPPER_RAMP_VELOCITY_START,
+	/** maximum ramp velocity, for controllers which provide ramping functionality */
+	STEPPER_RAMP_VELOCITY_MAX,
+	/** stop ramp velocity, for controllers which provide ramping functionality */
+	STEPPER_RAMP_VELOCITY_STOP,
+	/** threshold value for sensor-less stall detection */
+	STEPPER_STALL_DETECTION_THRESHOLD,
+	/** enable/disable stall detection */
+	STEPPER_STALL_DETECTION_STATUS,
+	/** retrieve the actual load detected by motor controller */
+	STEPPER_LOAD_DETECTION,
+};
+
+/**
+ * @typedef stepper_enable_t
+ * @brief enable or disable the stepper motor controller.
+ */
+typedef int32_t (*stepper_enable_t)(const struct device *dev, bool enable);
+
+/**
+ * @typedef stepper_move_t
+ * @brief Stepper Motor Controller shall move the motor with given steps
+ */
+typedef int32_t (*stepper_move_t)(const struct device *dev, int32_t steps);
+
+/**
+ * @typedef stepper_set_velocity_t
+ * @brief Set the velocity in steps per seconds. For controllers such as DRV8825 where you
+ * toggle the STEP Pin, the pulse_length would have to be calculated based on this parameter in the
+ * driver. For controllers where velocity can be set, this parameter corresponds to max_velocity
+ */
+typedef int32_t (*stepper_set_velocity_t)(const struct device *dev, uint32_t steps_per_second);
+
+/**
+ * @typedef stepper_set_micro_step_res_t
+ * @brief Set the microstep resolution
+ */
+typedef int32_t (*stepper_set_micro_step_res_t)(const struct device *dev,
+						enum micro_step_resolution resolution);
+
+/**
+ * @typedef stepper_set_attr_t
+ * @brief Set the value of a certain attribute in stepper controller
+ */
+typedef int32_t (*stepper_set_attr_t)(const struct device *dev, enum stepper_attribute,
+				      const int32_t value);
```

Comment:
How should this affect the stepper in the case of setting target/actual position? Should it start moving immediately, or just update the reference frame?

Reply:
For most controllers, the stepper will start moving immediately if it is enabled and the velocity/acceleration parameters are not 0. I think that makes sense here too, but it needs to be commented.

Similar to @bjarki-trackunit's comments, I think these are better in separate methods. A comment like "setting the actual position may cause the motor to begin moving" would be appropriate in the method to update the actual position.